### ISSUE TITLE
In monorepo case: prepend ../../node_modules/.bin to $PATH

### ISSUE THREAD

**Issue by tunnckoCore** [pnpm/pnpm#1355]
### Expected behavior:

Say you have monorepo, and in its package.json scripts you have `eslint` or whatever executable, which is intentionally installed as devDependency on the root of the monorepo. 
So the bin is in the `monorepoRoot/node_modules/.bin` and not in the `monorepoRoot/packages/foo/node_modules/.bin`. So you are forced to prepend each executable with `../../node_modules/.bin/eslint` for example.

I know there is `recursive` command and `--filter`, but there are cases, where you will or need to be in the package folder and run scripts only from there.

### Actual behavior:

Errors, because can't find executable in the package folder node_modules/.bin

### Additional information:

 - `node -v` prints: v10.9
 - Windows, OS X, or Linux?: ArchLinux

**Comment by olstenlarck** [comment]
I would propose `pnpm exec` is good thing to have.
Since `pnpm` follows the npm convention of commands and etc, e.g. you can do `pnpm lint` as one may do with `yarn lint` which calls the `lint` script, OR if there is executable with that name.

So with that proposal we will be able to do `pnpm exec eslint src --fix` for example. In which case it will look for `./node_modules/.bin/eslint` if not found then it will assume it is monorepo, so search in the `../../node_modules/.bin`.

**Comment by olstenlarck** [comment]
Lets illustrate with separate cli what should happen, for example called `monora`.

Consider the following monorepo structure

```
monorepo/
- package.json
- node_modules/
- packages/
  + foo
    - package.json  
```

so two cases

1. we are in the `monorepo/packages/foo` and call `monora lint`
  + look at the package.json
  + if `private !== true` then prepend both `./node_modules/.bin` and `../../node_modules/.bin` to the `$PATH` variable
    - check if the `lint` script exist - execute it
    - otherwise search for `lint` executable in both same bin folders and call it
    - otherwise exit with non-zero code `ENOENT`
2. we are in to `monorepo/` and call `monora lint`
  + look at the package.json
  + if `private === true` then we are in the monorepo root, so prepend only the `./node_modules/.bin`
    - check if the `lint` script exist - execute it
    - otherwise search for `lint` executable in `./node_modules/.bin` and call it
    - otherwise exit with non-zero code `ENOENT`

**Comment by olstenlarck** [comment]
It's implementation is simple

the CLI 

```js
#!/usr/bin/env node

'use strict';

const path = require('path');
const proc = require('process');
const execa = require('execa');

const argv = proc.argv.slice(2);
const cwd = proc.cwd();
const pkgPath = path.resolve(path.join(cwd, 'package.json'));
const pkg = require(pkgPath); // eslint-disable-line global-require, import/no-dynamic-require

// TODO: support `pre` and `post` scripts hooks
// TODO: support API
const scriptName = argv.shift();

const cwdNodeBin = path.resolve(path.join('node_modules', '.bin'));
const upNodeBin = path.resolve(path.join('..', '..', 'node_modules', '.bin'));

proc.env.PATH = `${cwdNodeBin}:${proc.env.PATH}`;

const scriptCommand = pkg.scripts[scriptName];
const args = scriptCommand ? scriptCommand.split(' ') : argv;
const cmd = scriptCommand ? args.shift() : scriptName;

// we are in monorepo package cwd
if (pkg.private !== true) {
  proc.env.PATH = `${upNodeBin}:${proc.env.PATH}`;
}

execa(cmd, args, { env: proc.env, stdio: 'inherit' });

```

**Comment by olstenlarck** [comment]
Or probably implemented behind the `run`?

```
npm run eslint src --fix
# and in Pnpm
pnpm run eslint src --fix
# and Yarn (already partially possible)
yarn eslint src --fix
```

The case of Yarn: it already can run executables if they exists in `node_modules/.bin`, not work in the mmonorepo case.

I'm gonna propose PRs to all th managers. Or would it be better to unify it around new command name e.g. `npm exec`, `yarn exec` and `pnpm exec`?

Thoughts?

**Comment by olstenlarck** [comment]
Okey, fully working example and API is living at [olstenlarck/monora](https://github.com/olstenlarck/monorepo/tree/master/packages/monora) 

Tests are still missing, but that's only because it is tricky to implement - will need a bit more things to the api.

It does exactly what i explained in above comments with single dependency and 2 cool free bonuses: 1) defining scripts through "js config file" (optional), escaping the limitation of json, like `(npm|yarn|pnpm).scripts.js` which can be CJS or ESM; and 2) shareable presets.

The scripts/tasks in this config file are nothing more than just exporting a string const. Everything is exactly the same as in `pkg.scripts` territory, except it is in js file and ran through `execa`. 

How it can be used here to implement `pnpm exec` command? Pretty straightforward and easy.

```js
// `manager` option controls what the optional config file name can be
monora({ manager: 'pnpm' }).catch(() => process.exit(1))
```

**Comment by zkochan** [comment]
I think it would be better to go with a new command. `pnpm exec` seems like a good name. We already have a command called `pnpm recursive exec` that does a similar thing but without modifying PATH.

I suggest to make it just for running bins and ignore scripts. There is already `pnpm run` for running scripts.

**Comment by zkochan** [comment]
Regarding the root of the monorepo. pnpm does not require a private `package.json` file in the root of the repo. pnpm searches for a `pnpm-workspace.yaml` to find the root of the repo.

TBH, I don't like the way monorepos are designed at the moment. I understand that it is convenient to install tooling in the root. However, installing things in the root for repo means that every single package in the repo may have access to dependencies that are not declared in their `package.json`. It can result in bugs that will be detected only after publishing the packages

**Comment by olstenlarck** [comment]
> pnpm does not require a private package.json file in the root of the repo. pnpm searches for a pnpm-workspace.yaml to find the root of the repo.

Yea, i know. That's cool thing too.

> However, installing things in the root of the repo means that every single package in the repo may have access to dependencies that are not declared in their package.json.

I understand that, and that's one of the general principles of pnpm - to not have access to other deps other than those defined in package's package.json.
But in same time, installing tooling as devDependencies in the monorepo root, is one of the main things of monorepo features - others are management, installation and managing versions (synced or independent). But also, I think that this isn't so big and common problem. The common case is that every package is uses its depedencies. There is eslint rules/plugins, best practices and such things that will signal you if you are using something that isn't in the `dependencies`. And it should be that way and that's the correct way.

> I suggest to make it just for running bins and ignore scripts.

The js config file is just a bonus, a result fact from wrapping up the API interface. It's completely optional thing and could be disabled entirely.

No, you can't ignore scripts. Because the whole thing is that we can't run does scripts (CLI bins defined in those scripts), because the bins are actually installed in the monorepo root. So both `pnpm run` and `pnpm recursive run` don't work well.

Example for what i'm talking:
If you have installed `eslint` in the root and have

```json
{
  "scripts": "eslint src --fix"
}
```

then you can run it without problems. If above is in some package then you get error that `eslint` is not found, because it resides in the monorepo root node_modules/.bin and not in the package/node_modules/.bin folder.

What's currently possible with monora is the following

```

# if `pkg.scripts.cov` is 'nyc ava'
monora cov --tap
# already possible with Yarn
yarn cov --tap

# but also just running installed CLIs
monora ava --tap
monora eslint src --fix
# already possible with Yarn
yarn ava --tap
yarn eslint src --fix

# or existing platform commands
# (not possible with yarn)
monora ls -al
monora echo "foo bar"
```

What i want to propose to all the package managers is to have lets say `exec` command, which works in the common case (single-repo project) and also in the monorepo case. It will easier a lot of things.

If not command name that is same across the managers, then we are still good, because all that thing is implemented as separate package, so every manager can decide behind what command to be represented:

- In the @npm case it make sense to be behind the `npm run`
- In the @yarnpkg it make sense to be behind `yarn` - since it already works mostly that way 
- In pnpm case make sense to be behind `pnpm run` too, but `pnpm exec` would be good too.

All that said, even if not implemented anywhere, there is one handy cool trick/workaround that can be used - all package mangers has the `start` command, so you can define pkg.scripts.start and execute/manage everything through that start command. The @deepsweet's [Start](https://github.com/deepsweet/start) already takes advantage of this - so whatever the package manger you are using you can `yarn start dev`, `yarn start cov` or `npm start dev` and `npm start test`. :)

**Comment by zkochan** [comment]
> No, you can't ignore scripts. Because the whole thing is that we can't run does scripts (CLI bins defined in those scripts), because the bins are actually installed in the monorepo root. So both pnpm run and pnpm recursive run don't work well.

That is not what I meant. I am OK to extend the PATH for scripts. So it is fine that you run `pnpm run lint` in a package and the script will find eslint in the root of the repo.

What I meant was, I think it would be smarter to keep using `pnpm run` for running scripts and the new command would only execute commands. So to run a script you would be able to run `pnpm run lint` but would not be able to run `pnpm exec lint`. `pnpm exec eslint` would work though (or `pnpm exec pnpm run lint`)

**Comment by zkochan** [comment]
> But in same time, installing tooling as devDependencies in the monorepo root, is one of the main things of monorepo features - others are management, installation and managing versions (synced or independent). But also, I think that this isn't so big and common problem. The common case is that every package is only using its depedencies and in its test its devDependencies. There is eslint rules/plugins, best practices and such things that will signal you if you are using something that isn't in the dependencies. And it should be that way and that's the correct way.

In theory, you could install all the toolings in a subfolder and then use a custom bin location.

So you can have a package.json in `monorepo/tooling/package.json`. It would contain eslint and everything you need.

Then you can have a `.npmrc` there which sets the bin location to `../node_modules/.bin`. So dependencies will be installed inside `monorepo/tooling/node_modules` but all the bins will be linked into `monorepo/node_modules/.bin`.

Scripts have access to all the commands. App code has no access to modules.

**Comment by zkochan** [comment]
PR is in progress #1870

**Comment by zkochan** [comment]
:ship: 3.5.0-3

**Comment by zkochan** [comment]
🚢 3.5.0

**Comment by iampeterbanjo** [comment]
i have version 3.6.5 installed and i'm looking for a way to configure what used to be a yarn workspace into a pnpm workspace. I want to run `pnpm run lint` from the root of my monorepo and have it execute `pnpx eslint --fix-dry-run ./packages/e2e ./packages/server` but i'm getting module resolution errors everywhere e.g. `1:1   error  Resolve error: unable to load resolver "node"   import/no-unresolved`.

does this feature address my concerns?

> So to run a script you would be able to run pnpm run lint but would not be able to run pnpm exec lint. pnpm exec eslint would work though (or pnpm exec pnpm run lint)

What's the recommended way to setup my monorepo as you suggest? there doesn't seem to be documentation on `pnpm exec`

**Comment by zkochan** [comment]
This seems like the eslint issue. Eslint uses some dependencies that are not declared as deps. Try setting `shamefully-flatten=true` in the `.npmrc` in the root of the repo and run `pnpm multi install --force`.

**Comment by iampeterbanjo** [comment]
Thanks. i tried that and it worked for eslint.  But I have a problem with the dynamic loading of dependencies - I'll open a separate issue
